[ellie.klm 김예지] 지하철 노선도 step3 코드리뷰 요청드립니다.

Ye-Ji-Kim opened the pull request:
안녕하세요 ellie 입니다.
지하철 노선도 미션 리뷰 부탁드립니다 :)
 
[질문]
서비스를 만들어 도메인과 분리하고자 하였는데, 도메인 보다는 서비스에 로직이 많아졌습니다.
서비스에 있는 로직을 도메인으로 옮기기 위한 팁을 여쭤보고 싶습니다.

Ye-Ji-Kim committed:
feat: CRUD 구현

Station CRUD, Line CRUD 구현

Ye-Ji-Kim committed:
feat: LineResponse default 생성자 추가

Ye-Ji-Kim committed:
feat: step2 지하철 구간 추가 API구현

tttuer committed:
feat: 1단계 실습

tttuer committed:
Merge branch 'step1' of https://github.com/Ye-Ji-Kim/spring-subway-admin-kakao into step1

Ye-Ji-Kim committed:
feat: 구간 등록 API 에러 수정

tttuer committed:
Merge branch 'step1' of https://github.com/Ye-Ji-Kim/spring-subway-admin-kakao into step1

Ye-Ji-Kim committed:
feat: step2 지하철 구간 추가 API 최종 수정

Ye-Ji-Kim committed:
feat: step3 jdbc로 변경

tttuer committed:
Merge branch 'step1' of https://github.com/Ye-Ji-Kim/spring-subway-admin-kakao into step1

tttuer committed:
feat: 버그 수정

Test실행 했을때 오류 수정

Ye-Ji-Kim committed:
feat: step3 구간 추가 API 에러 수정 및 구간 삭제 API 구현

Ye-Ji-Kim committed:
feat: 지하철 구간 삭제 에러 수정

friediceday commented on src/main/java/subway/line/Section.java:
```diff
@@ -0,0 +1,80 @@
+
+import java.util.List;
+
+public class Section {
```
섹션은 다른 라인, 스테이션 패키지와 마찬가지로 별도로 분리하면 어떨까요

friediceday commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,96 @@
+
+@RestController
+public class LineController {
+    private LineDao lineDao;
```
컨트롤러가 dao를 직접 접근하기보단 서비스가 해주는게 좋을 것 같습니다.

friediceday commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    @PostMapping("/lines")
+    public ResponseEntity<LineResponse> createStation(@RequestBody LineRequest lineRequest) {
```
url은 lines인데 메서드명은 station인데 어떤게 맞나요?

friediceday commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,96 @@
+        if (newLine == null) {
+            return ResponseEntity.badRequest().build();
+        }
+        sectionDao.insert(new Section(newLine.getId(), lineRequest.getUpStationId(), lineRequest.getDownStationId(), lineRequest.getDistance()));
```
실제로 없는 stationid가 들어오면 어떻게 되나요?

Ye-Ji-Kim replied:
프론트를 실행하여 브라우저로 확인해보았을 때,  노선 생성 시 상행과 하행은 기존에 등록된 station 중 선택되도록 되어있었습니다.
따라서 없는 stationId 가 들어오는 경우는 없을 것이라 생각하고 코드를 작성하였습니다.
이 경우에도 예외처리가 필요한지, 그렇다면 어떻게 처리하면 좋을지(그냥 badrequest를 status code로 보내면 될까요?) 여쭤보고 싶습니다!

friediceday replied:
프론트에서 정상적인 경우 가능하지 않지만, 이상한 요청을 보내는게 불가능하진 않습니다. (버그, 혹은 악의적으로)
따라서 이런 엣지 케이스에 대한 처리가 있으면 더 좋겠습니다.
그리고 클라이언트에서 비정상 요청을 한 것이므로 말씀하신대로 그냥 에러 응답을 내주면 괜찮을 것 같습니다.

friediceday commented on src/main/java/subway/line/LineService.java:
```diff
@@ -0,0 +1,34 @@
+        this.sectionService = sectionService;
+    }
+
+    public List<Station> getStations(Line line) {
```
해당 라인의 스테이션들을 갖고오기 위해서,, 해당 라인의 섹션을 모두 가져온 다음 스테이션id를 추출하고 각각의 스테이션id로 다시 스테이션을 쿼리하고 있습니다.
N+1 쿼리라고 하고요, 비효율적입니다.
스테이션에 lineid가 있으면 아주 간단하게 가능할 것 같은데요.

Ye-Ji-Kim replied:
노선의 스테이션들을 리턴해줄 때, 상행 종점부터 하행 종점까지의 역들을 순서대로 정렬하여 보내주어야 한다고 이해하였습니다.
따라서 이를 위해 섹션들을 순서대로 가져온 후 스테이션id의 순서를 추출하여 스테이션을 쿼리하였습니다.
고민해보았지만 스테이션에 lineid를 추가하는 것만으로는 섹션을 활용하지 않고 역의 순서를 고려한 스테이션 쿼리를 할 방도를 생각해내지 못하였습니다. 이에 대해 조금 더 조언을 구하고 싶습니다.

friediceday replied:
디비 쿼리와 정렬을 분리해서 생각해보면 어떨까요?

friediceday commented on src/main/java/subway/line/SectionService.java:
```diff
@@ -0,0 +1,120 @@
+                .collect(Collectors.toMap(Section::getUpStationId, Section::getDownStationId));
+        Map<Long, Integer> downStationAndDistance = sectionDao.findByLineId(lineId).stream()
+                .collect(Collectors.toMap(Section::getDownStationId, Section::getDistance));
+
```
쿼리를 한번만 해도 되지 않을까요

friediceday commented on src/main/java/subway/line/SectionService.java:
```diff
@@ -0,0 +1,120 @@
+                    downStationAndDistance.get(upStationAndDownStation.get(startId))));
+            startId = upStationAndDownStation.get(startId);
+        }
+        return result;
```
혹시 이렇게 순회하면서 찾는 이유는 디비에 있는 섹션이 잘못된게 있을까봐 그런건가요?

friediceday commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,96 @@
+
+    @DeleteMapping("/lines/{id}")
+    public ResponseEntity deleteStation(@PathVariable Long id) {
+        lineDao.deleteById(id);
```
이렇게 단 한줄이라도 dao를 직접 접근하지 않고 서비스만 접근하면 좋겠습니다.

friediceday commented on src/main/java/subway/line/LineDao.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    public List<Line> findAll() {
+        String sql = "select id, name, color from line";
```
이렇게 여러개를 리턴할 경우 항상 리밋이 있어야 안전할 것 같습니다.

friediceday commented on src/main/java/subway/line/SectionService.java:
```diff
@@ -0,0 +1,120 @@
+        sectionDao.insert(section);
+    }
+
+    private void updateWhenUpStationMatching(Section section) {
```
아래와 유사한데 하나로 합칠 수 있을까요

friediceday commented on src/main/java/subway/line/SectionService.java:
```diff
@@ -0,0 +1,120 @@
+    public int delete(Long lineId, Long stationId) {
+        List<Section> sections = sectionDao.findByLineId(lineId);
+        Section upStationMatch = sections.stream()
+                .filter(section1 -> section1.getUpStationId() == stationId)
```
1이라는 변수명은 안쓰는게 좋을 것 같습니다

friediceday commented on src/main/java/subway/line/SectionService.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    public int delete(Long lineId, Long stationId) {
+        List<Section> sections = sectionDao.findByLineId(lineId);
```
line에 head, tail 스테이션이 누군지 알면 그걸 이용하면 섹션도 찾을 수 있지 않을까요?

그런데 ui도 그렇고, LineRequest를 봐도 노선 생성시 head, tail 스테이션을 넣게 되어있는 것 같은데 디비 스키마엔 없네요.

friediceday commented on src/main/java/subway/line/SectionService.java:
```diff
@@ -0,0 +1,120 @@
+    public void insert(Section section) {
+        section.checkValidInsert(showAll(section.getLineId()));
+
+        Map<Long, Long> upStationAndDownStation = sectionDao.findByLineId(section.getLineId()).stream()
```
여기도 아마 line에 head, tail 정보가 있다면 모두 가져오지 않아도 될 것 같아요.

friediceday commented on src/main/java/subway/line/LineDao.java:
```diff
@@ -0,0 +1,73 @@
+        return jdbcTemplate.queryForObject(sql, lineRowMapper, id);
+    }
+
+    public List<Line> findLineByName(String name) {
```
find lines... 는 어떨까요

friediceday commented on src/main/java/subway/station/StationController.java:
```diff
@@ -6,25 +6,39 @@
 
 @RestController
 public class StationController {
+    private StationDao stationDao;
```
여기서도 마찬가지로 dao를 직접 접근하는게 아니라 로직은 모두 서비스에 있다고 생각하시면 좋을 것 같아요

ljh131 reviewed (changes requested):
안녕하세요. 테스트를 많이 추가 해보시면 좋겠습니다. 특히 2단계에 나와있는 구간 등록/제거에 관련된 모든 정상적인 케이스, 그리고 가능하시면 엣지 케이스도 추가해보시면 좋겠습니다.
질문하신건 따로 답글 달아드리겠습니다.

ljh131 commented:
> [질문]
> 서비스를 만들어 도메인과 분리하고자 하였는데, 도메인 보다는 서비스에 로직이 많아졌습니다.
> 서비스에 있는 로직을 도메인으로 옮기기 위한 팁을 여쭤보고 싶습니다.

말씀하시는 도메인이라는게 구체적으로 어떤건가요?

Ye-Ji-Kim committed:
refactor: section 관련 클래스들 패키지로 분리

Ye-Ji-Kim committed:
refactor: 노선 생성 API  수정, Line 객체와 DB schema 수정

Ye-Ji-Kim committed:
refactor: 노선 조회 API 수정 및 관련 테스트 보강

Ye-Ji-Kim committed:
refactor: 노선 수정 및 삭제 API 수정, 구간관련 컨트롤러 SectionController로 이동

Ye-Ji-Kim committed:
refactor: stationDao 접근 로직 StationService로 이동, 테스트 추가

Ye-Ji-Kim committed:
refactor: 구간 추가 및 제거 로직 수정, 테스트 추가

Ye-Ji-Kim committed:
refactor: code reformat

Ye-Ji-Kim commented:
조언 반영하여 새로 수정하였습니다!!

[질문]
1. 질문드렸던 도메인은 Line, Section과 같은 객체입니다.
강의에서 Line, Section 과 같은 도메인 객체에 최대한 로직을 담고 서비스의 로직을 최소화하라고 하셨는데,
Dao 접근이 빈번하다보니 서비스에 대부분의 로직을 담게된 것 같습니다ㅠㅠ

2. 피드백 주셨던 station에 lineId를 추가하여 쿼리하는 방법을 적용해보고자 했으나, 환승구간의 경우 한 station이 여러 lineId를 가지게 되어 lineId에 따른 중복된 station을 저장하게 되는 것을 발견했습니다. 우선 기존의 방법을 그대로 두었지만 이 경우 중복된 데이터를 저장하는 것이 쿼리를 여러번 수행하는 것보다 좋은 방법인지 궁금합니다.

3. **Dao는 **Service에서만 접근할 수 있도록 구현하고자 하였으나, 그러다보니 lineService와 SectionService 사이에 순환 종속성 문제가 발생하였습니다. 검색해보니 순환 종속성은 최대한 지양하는 것이 좋다고 하여 LineService에서 SectionDao를 사용하도록 하였습니다. 이것도 괜찮은 해결 방법인지, 더 좋은 방법이 있는지 여쭤보고 싶습니다.

friediceday commented on src/main/java/subway/line/LineDao.java:
```diff
@@ -0,0 +1,81 @@
+        return jdbcTemplate.query(sql, lineRowMapper);
+    }
+
+    public Line findLineById(Long id) {
```
primitive를 사용하지 않고 전체적으로 reference를 사용하셨는데 이유가 있을까요?

friediceday commented on src/main/java/subway/station/StationDao.java:
```diff
@@ -1,33 +1,70 @@
-        Station persistStation = createNewObject(station);
-        stations.add(persistStation);
-        return persistStation;
+        if (countByName(station.getName()) > 0) {
```
존재하는지 여부만 알면 되지 않을까요

friediceday commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,69 @@
+        try {
+            Line newLine = lineService.createLine(line);
+            LineResponse lineResponse = new LineResponse(newLine.getId(), newLine.getName(), newLine.getColor());
+            return ResponseEntity.created(URI.create("/lines/" + newLine.getId())).body(lineResponse);
```
그냥 ok로 해도 될텐데 created로 하신 이유가 있나요

friediceday commented on src/main/java/subway/line/LineDao.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    public List<Line> findAll() {
+        String sql = "select id, name, color, up_station_id, down_station_id from line limit 50";
```
상수를 사용하면 좋겠습니다. 그리고 50이면 충분한가요?

friediceday commented on src/main/java/subway/line/LineService.java:
```diff
@@ -0,0 +1,83 @@
+            section = sections.get(iterStationId);
+            result.add(section);
+            iterStationId = section.getDownStationId();
+        } while (iterStationId != tailStationId);
```
한 가지 더 있음직한 문제를 소개해드리겠습니다. 
만약 데이터가 꼬여있어서 중간에 circular 형태가 있다면 loop이 끝나지 않을 것입니다. 
어떻게 하는게 좋을까요?

friediceday commented on src/main/java/subway/section/SectionService.java:
```diff
@@ -0,0 +1,82 @@
+        if (line.getUpStationId() == section.getDownStationId()) { // 상행 종점으로 삽입
+            sectionDao.insert(section);
+            lineService.updateLine(new Line(line.getId(), line.getName(), line.getColor(), section.getUpStationId(), line.getDownStationId()));
+            return;
```
이렇게 중간중간에 리턴하지 않도록 하면 어떨까요

friediceday commented on src/main/java/subway/section/SectionService.java:
```diff
@@ -0,0 +1,82 @@
+            sectionDao.insert(section);
+            int newDistance = currentSection.getDistance() - section.getDistance();
+            sectionDao.update(new Section(currentSection.getId(), section.getLineId(), section.getDownStationId(), currentSection.getDownStationId(), newDistance));
+            return;
```
아래와 중복인 부분이 꽤 있는 것 같습니다. 공통으로 빼면 어떨까요

friediceday commented on src/main/java/subway/section/SectionService.java:
```diff
@@ -0,0 +1,82 @@
+        this.lineService = lineService;
+    }
+
+    public void insertNewSection(Section section) {
```
메서드가 전체적으로 긴 것 같습니다.

friediceday commented on src/main/java/subway/section/SectionService.java:
```diff
@@ -0,0 +1,82 @@
+
+    }
+
+    public int deleteSection(Long lineId, Long stationId) {
```
이 메서드도 좀 긴 것 같습니다.

friediceday commented on src/main/java/subway/section/SectionService.java:
```diff
@@ -0,0 +1,82 @@
+        sectionDao.insert(newSection);
+        deleteCount += sectionDao.delete(upSection);
+        deleteCount += sectionDao.delete(downSection);
+        return deleteCount;
```
사용하지 않는 값이면 굳이 안넘겨도 되지 않을까 싶습니다.

friediceday commented on src/test/java/subway/line/SectionAcceptanceTest.java:
```diff
@@ -36,36 +38,67 @@ public void setUp() {
     }
 
-    @DisplayName("지하철 구간을 등록한다.")
+    @DisplayName("기존 구간 사이에 구간 등")
```
오타가 있네요

ljh131 reviewed (changes requested):
잘 수정하셨습니다. 전반적으로 좋아진 것 같습니다.
조금 더 개선할 수 있을 부분들은 커멘트 드렸습니다. 질문이나 의견 있으시면 남겨주세요.

ljh131 commented:
질문드렸던 도메인은 Line, Section과 같은 객체입니다.
강의에서 Line, Section 과 같은 도메인 객체에 최대한 로직을 담고 서비스의 로직을 최소화하라고 하셨는데,
Dao 접근이 빈번하다보니 서비스에 대부분의 로직을 담게된 것 같습니다ㅠㅠ
-> 이 부분은 개인 성향으로 주관적이라고 생각합니다.

피드백 주셨던 station에 lineId를 추가하여 쿼리하는 방법을 적용해보고자 했으나, 환승구간의 경우 한 station이 여러 lineId를 가지게 되어 lineId에 따른 중복된 station을 저장하게 되는 것을 발견했습니다. 우선 기존의 방법을 그대로 두었지만 이 경우 중복된 데이터를 저장하는 것이 쿼리를 여러번 수행하는 것보다 좋은 방법인지 궁금합니다.
-> 말씀하신대로 역이 여러 라인을 가질 수 있겠군요. 그렇다면 더 복잡하게 할 필요는 없을 것 같습니다. 이미 지금과 같이 수정하신 방법도 좋아보입니다.

**Dao는 **Service에서만 접근할 수 있도록 구현하고자 하였으나, 그러다보니 lineService와 SectionService 사이에 순환 종속성 문제가 발생하였습니다. 검색해보니 순환 종속성은 최대한 지양하는 것이 좋다고 하여 LineService에서 SectionDao를 사용하도록 하였습니다. 이것도 괜찮은 해결 방법인지, 더 좋은 방법이 있는지 여쭤보고 싶습니다.
-> 네 순환 참조는 없애는게 바람직하다고 봅니다. 그리고 이미 두 객체가 서로 많이 연관되어 있기 때문에 다른 dao를 접근해도 되지 않을까 싶습니다.

Ye-Ji-Kim committed:
refactor: Dao 삽입시 데이터 존재여부 확인 로직 수정

Ye-Ji-Kim committed:
feat: 기능 요구사항 상세 추가

Ye-Ji-Kim committed:
refactor: 모든 데이터 조회 쿼리와 데이터 정렬 로직 수정

- LineDao, StationDao의 findAll 쿼리시 limit 값 조정
- LineService의 getSortedSections 메소드에서 정렬 로직 수정하여 무한루프 가능성 방지

Ye-Ji-Kim committed:
refactor: 구간 추가 로직 수정

Ye-Ji-Kim committed:
refactor: 구간 삭제 로직 수정

Ye-Ji-Kim commented:
다른 미션을 수행하느라 수정이 늦었습니다ㅠㅠ
피드백 주신 부분 모두 고민하여 수정하였고, 질문 주신 사항에 대한 답변 및 추가 질문은 아래에 정리하였습니다.

- primitive를 사용하지 않고 전체적으로 reference를 사용하셨는데 이유가 있을까요?
-> 기존 베이스 코드가 그렇게 구현되어 있어서 통일성을 유지하기 위해 저도 reference를 사용하였습니다. Reference 보다는 primitive를 사용하는 것이 더 바람직한가요?

- 그냥 ok로 해도 될텐데 created로 하신 이유가 있나요
-> 201 코드를 반환하도록 기능 요구사항이 되어있어서 created를 사용하였습니다. 기능 요구사항 상세는 requirements.md 파일로 추가하였습니다.

- 상수를 사용하면 좋겠습니다. 그리고 50이면 충분한가요?
-> 서울 지하철 노선 수 23개, 역 수 324개를 바탕으로 line은 50, station은 500으로 limit을 조정하였습니다. 이렇게 limit을 결정하는 것이 맞는 지 모르겠습니다.

- 한 가지 더 있음직한 문제를 소개해드리겠습니다.
만약 데이터가 꼬여있어서 중간에 circular 형태가 있다면 loop이 끝나지 않을 것입니다.
어떻게 하는게 좋을까요?
-> sections의 사이즈만큼 루프를 돌도록 수정하였습니다.

friediceday commented on src/main/java/subway/section/SectionService.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    public void deleteSection(Long lineId, Long stationId) {
+        if (sectionDao.countByLineId(lineId) < 2) {
```
상수를 사용하면 어떨까요

friediceday commented on src/main/java/subway/section/SectionService.java:
```diff
@@ -0,0 +1,117 @@
+        if(!isUpdated){
+            throw new IllegalArgumentException("상행역과 하행역 모두 노선에 등록되어 있지 않아 구간을 추가할 수 없습니다.");
+        }
+
```
모두 실패할때만 실패하게 되어있는데요, 일부 실패하는건 괜찮나요?

friediceday commented on src/main/java/subway/station/StationController.java:
```diff
@@ -6,25 +6,40 @@
+        } catch (Exception e) {
+            return ResponseEntity.badRequest().build();
+        }
     }
```
조금 더 예외처리를 상세히 해볼 수 있을 것 같아요. 버그로 인해서 익셉션이 발생한 경우에는 bad req라고 볼 수 없으니까요.

friediceday commented on src/main/java/subway/section/SectionService.java:
```diff
@@ -0,0 +1,117 @@
+            Section currentSection = sectionDao.findByLineIdAndUpStationId(line.getId(), section.getUpStationId());
+            int newDistance = checkValidAndGetDistance(section, currentSection);
+            sectionDao.update(new Section(currentSection.getId(), section.getLineId(), section.getDownStationId(), currentSection.getDownStationId(), newDistance));
+            return true;
```
이런 부분은 아래와 중복되는 부분이 꽤 있으니 메서드로 빼는 것도 괜찮을 것 같습니다

ljh131 reviewed (commented):
수고하셨습니다. 잘 수정해주셨습니다. 이번 단계는 여기서 마무리 하겠습니다. 
사소한 커멘트 드렸으니 참고해보시면 좋겠습니다.

그리고 되도록 primitive 사용을 권장드립니다. 아래에 많은 답변이 있으니 읽어보시고 도움 되시면 좋겠습니다.
https://stackoverflow.com/questions/5199359/why-do-people-still-use-primitive-types-in-java
